src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
+                                {
+                                    BasicBlock*  testPred       = predIter->flBlock;
+                                    unsigned int predPosNum     = positionNum(testPred);
+                                    unsigned int blockPosNum    = positionNum(block);
+                                    unsigned int lastExitPosNum = positionNum(lastExitBlock);
+
+                                    if (loopBlocks.isMember(predPosNum) ||
+                                        ((predPosNum >= blockPosNum) && (predPosNum <= lastExitPosNum)))
+                                    {
+                                        // This pred is in the loop (or what will be the loop if we determine this
+                                        // run of exit blocks doesn't include a side-entry).
+
+                                        if (predPosNum < blockPosNum)
+                                        {
+                                            // We've already counted this block as an exit, so decrement the count.
+                                            --exitCount;
+                                            // It's ok to leave `exit` alone even if it points to testPred; this
+                                            // range of blocks isn't in the loop so it must include an exit, which
+                                            // we'll update `exit` to when we visit it.
+                                        }
+                                    }
+                                    else
+                                    {
+                                        // This pred is not in the loop, so this constitutes a side-entry.
+                                        goto NO_LOOP;
+                                    }
+                                }
+
+                                // Either we're going to abort the loop on a subsequent testBlock, or this
+                                // testBlock is part of the loop.
+                                loopBlocks.insert(testBlock->bbNum);
+                            }
+
+                            // Now that we've marked these blocks as part of the loop, iterate without updating
+                            // `previous` so that we'll analyze them as part of the loop.
+                            continue;
+                        }
+
+                        fgUnlinkRange(block, lastExitBlock);
+                        fgMoveBlocksAfter(block, lastExitBlock, moveAfter);
+                        ehUpdateLastBlocks(moveAfter, lastExitBlock);
+
+                        // If lastExitBlock fell through to nextLoopBlock, then it also would have been in the loop.
+                        assert(!lastExitBlock->bbFallsThrough());
+
+                        if (moveAfter->bbFallsThrough())
+                        {
+                            // We've just inserted blocks between moveAfter and moveBefore which it was supposed
```

Comment:
Would be clearer to me if `moveBefore` was actually a thing here. It took me a while to figure out that `lastExitBlock->bbNext` was the block that `moveAfter` falls through to originally.